### ISSUE TITLE
Generated PKs are ignored in model query methods

### ISSUE THREAD

**Issue by daddz** [stephenafamo/bob#83]
I really like the idea of the generated setters especially when inserting new rows based on requests that may or may not contain some data.

But for updating a row there does not seem to be an easy way to use them - at least not one that I found so far.

It would be nice if something like this could work:
```
foo, _ := models.FindFoo(ctx, db, id)

setter := &models.FooSetter{}

// fill out the setter depending on the received request
if req.bar {
  setter.Bar.Set(req.bar)
}
// and so on

foo.Update(ctx, db, setter)
```

Did I overlook something in the docs/code or is my approach flawed anyway?

**Comment by daddz** [comment]
I have tried to do it like this but the generated SQL is missing the PK.

```
foo, _ := models.FindFoo(ctx, db, id)

setter := &models.FooSetter{}
setter.Bar.Set("foobar")

models.FooTable.UpdateMany(ctx, s.db, setter, foo)
```

Generated SQL:
```
UPDATE "foo" AS "foo" SET
"bar" = $1
WHERE ( IN ())
```

**Comment by stephenafamo** [comment]
> I have tried to do it like this but the generated SQL is missing the PK.

This should work. Can you make sure that no error was returned from `FindFoo`?

**Comment by daddz** [comment]
> > I have tried to do it like this but the generated SQL is missing the PK.
> 
> This should work. Can you make sure that no error was returned from `FindFoo`?

In the actual code are error checks and it returns correctly without error.

It seems like the following call in `table.go` does not contain any pks after running:
`pks, pkVals, err := internal.GetColumnValues(t.mapping, t.mapping.PKs, rows...)`

Values after it executed (via debugger):
```
pks = {[]string} len:0, cap:8
pkVals = {[][]bob.Expression} len:1, cap:1
╚ 0 = {[]bob.Expression} len:0, cap:8
```

**Comment by daddz** [comment]
> > But for updating a row there does not seem to be an easy way to use them - at least not one that I found so far.
> 
> This is among the changes in #81, it seems obvious that Setters should be able to work as an `UpdateMod`

Could you give an example how the setter would be used together with the auto-generated Update method?

Neither the `models.FooTable.Update` nor the `models.Foo.Update` accept a Setter/UpdateMod.

**Comment by stephenafamo** [comment]
> Here's a minimal working example for the UpdateMany issue:

I have been unable to replicate this. In my own setup, the PKs were picked up correctly.


> Could you give an example how the setter would be used together with the auto-generated Update method?

It would be for a new method `FooTable.UpdateAll`. It would also be able to be used as an update query mod whenever an update query is being built.

**Comment by daddz** [comment]
> I have been unable to replicate this. In my own setup, the PKs were picked up correctly.

I think I found the culprit: by using the `generated` declaration for the `id` column (in sql and therefore the generated struct) it ends up getting filtered out in https://github.com/stephenafamo/bob/blob/main/internal/reflect.go#L186

This causes it to not find the PK properly in my case. Is this intended behavior?

> It would be for a new method `FooTable.UpdateAll`. It would also be able to be used as an update query mod whenever an update query is being built.

Sounds great!

**Comment by stephenafamo** [comment]
> I think I found the culprit: by using the `generated` declaration for the `id` column (in sql and therefore the generated struct) it ends up getting filtered out

This is a bug. I'll try to fix as soon as I can

**Comment by stephenafamo** [comment]
Should have been fixed by c82ba2d
Can you try it and let me know?

**Comment by daddz** [comment]
> Should have been fixed by [c82ba2d](https://github.com/stephenafamo/bob/commit/c82ba2d7b6bed22c627cf31d638edda13ac7ca01) Can you try it and let me know?

Can confirm. It's working properly with that commit!
